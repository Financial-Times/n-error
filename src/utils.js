export const removeObjectKeys = obj => keys => {
	const output = obj;
	if (Array.isArray(keys)) {
		Object.keys(obj).forEach(key => {
			if (keys.includes(key)) {
				delete output[key];
			}
		});
		// if (obj instanceof Error) {
		// 	['message', 'stack', 'code'].forEach(key => {
		// 		if (keys.includes(key)) {
		// 			Object.assign(obj, { [key]: undefined });
		// 		}
		// 	});
		// }
		return output;
	} else if (typeof keys === 'string' && keys !== '') {
		const k = keys;
		Object.keys(obj).forEach(key => {
			if (key === k) {
				delete output[key];
			}
		});
		// if (obj instanceof Error && ['message', 'stack', 'code'].includes(k)) {
		// 	Object.assign(output, { [k]: undefined });
		// }
		return output;
	}
	throw Error('keys need to be formatted in [string] or non-empty string');
};

export default {
	removeObjectKeys,
};
