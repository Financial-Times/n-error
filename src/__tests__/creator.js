import nError, { ERROR_STATUS_TEXT_MAP } from '../creator';
import { assertError } from '../utils';

describe('ERROR_STATUS_TEXT_MAP', () => {
	it('accessible via key name as number', () => {
		const type = ERROR_STATUS_TEXT_MAP[404];
		expect(type).toBe('NOT_FOUND');
	});

	it('accessible via key name as string', () => {
		const type = ERROR_STATUS_TEXT_MAP['404'];
		expect(type).toBe('NOT_FOUND');
	});
});

describe('nError', () => {
	describe('has methods', () => {
		it('to create error with status named after error type in camelCase', () => {
			const e = nError.notFound({ message: 'some error message' });
			expect(e.status).toBe(404);
			expect(e).toMatchSnapshot();
		});

		it('for all status defined in ERROR_STATUS_TEXT_MAP', () => {
			const methods = Object.keys(nError);
			expect(methods).toHaveLength(Object.keys(ERROR_STATUS_TEXT_MAP).length);
		});
	});

	describe('create NError', () => {
		it('create an Error with extended fields correctly', () => {
			const e = nError({ message: 'some message', foo: 'bar' });
			assertError(e);
		});

		it('create an ExtendedError with extend()', () => {
			const e = nError({ message: 'some message' });
			e.extend({ next: 'TEST' });
			expect(e.next).toBe('TEST');
		});

		it('create an ExtendedError with toUser()', () => {
			const e = nError({ message: 'some message' });
			e.toUser({ message: 'some other message' });
			expect(e.user).toEqual({ message: 'some other message' });
		});

		it('create an ExtendedError with setHandler()', () => {
			const e = nError({ message: 'some message' });
			e.setHandler('REDIRECT_TO_INDEX');
			expect(e.handler).toBe('REDIRECT_TO_INDEX');
		});

		// it('maintains native Error fields with rest spread', () => {
		// 	const e = new NError({ message: 'some message' });
		// 	const ee = new NError({ ...e, action: 'SOME_ACTION_TYPE' });
		// 	expect(ee.message).toBe('some message');
		// });
	});
});
