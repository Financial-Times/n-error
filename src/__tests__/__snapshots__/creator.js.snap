// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nError create NError create an Error with extended fields correctly 1`] = `
Object {
  "extend": [Function],
  "foo": "bar",
  "message": "some message",
  "setHandler": [Function],
  "toUser": [Function],
}
`;

exports[`nError has methods to create error with status named after error type in camelCase 1`] = `[Error: some error message]`;
